stdlib/universe/holt_winters/neldermead.go

```diff
@@ -0,0 +1,253 @@
+package holt_winters
+
+import "math"
+
+const (
+	defaultMaxIterations = 1000
+	// reflection coefficient
+	defaultAlpha = 1.0
+	// contraction coefficient
+	defaultBeta = 0.5
+	// expansion coefficient
+	defaultGamma = 2.0
+)
+
+// Optimizer is an implementation of the Nelder-Mead optimization method.
+// Based on work by Michael F. Hutt: http://www.mikehutt.com/neldermead.html
+type Optimizer struct {
+	// Maximum number of iterations.
+	MaxIterations int
+	// Reflection coefficient.
+	Alpha,
+	// Contraction coefficient.
+	Beta,
+	// Expansion coefficient.
+	Gamma float64
+
+	tracker *memoryTracker
+}
+
+// NewOptimizer returns a new instance of Optimizer with all values set to the defaults.
+func NewOptimizer(tracker *memoryTracker) *Optimizer {
+	return &Optimizer{
+		MaxIterations: defaultMaxIterations,
+		Alpha:         defaultAlpha,
+		Beta:          defaultBeta,
+		Gamma:         defaultGamma,
+		tracker:       tracker,
+	}
+}
+
+// Optimize applies the Nelder-Mead simplex method with the Optimizer's settings.
+func (o *Optimizer) Optimize(
+	objfunc func([]float64) float64,
+	start []float64,
+	epsilon,
+	scale float64,
+) (float64, []float64) {
+	n := len(start)
```

Comment:
I can see why it is hard to use Arrow here and why it makes sense to just track the allocations out of band from the slices.

Since we need mutability here but do not need the interoperability that Arrow gives us here should we just create a package to encapsulate this behavior? As seen above it is easy to have a typo and track a different value from what is actually allocated. Instead if you just asked for an allocated slice of floats and the tracking was internal to that call do errors could be made:

Maybe a simple API like this:

```
type SliceAllocator interface {
    AllocateFloats(n int) (data []float, free func() )
}
```

And used like:

```
v, free := o.allocator.Floats(n)
defer free()
```

Thoughts?

Reply:
Also like @aanthony1243  says `n` is small here i.e. likely always less than `100` so maybe we just don't care about tracking this smaller bit of memory usage? But it is `n^2` memory usage...